### PR TITLE
feat: finish migration of `/forums/recent-posts` to Inertia+React

### PR DESCRIPTION
**Pull request by wescopeland**
This PR is stacked on #2502 and has no functional changes. The purpose of this PR is to replace the Blade-powered recent posts page with the React-powered one.

**How to test this PR:**
* `sail artisan migrate` - the new recent posts Eloquent ORM query needs an index added.
* `sail npm run build` followed by `sail artisan inertia:start-ssr`.
* Navigate to `/forums/recent-posts`.

Summary of changes:
* Resource/Data classes are now used for automatic props mapping with `spatie/laravel-data`.
* TypeScript types are now auto-generated with `spatie/typescript-transformer`.
* The legacy `getRecentForumTopics()` helper has been removed.
* The controller code has moved to `ForumTopicController::recentlyActive()`, and an Eloquent ORM query is now utilized, along with `LengthAwarePaginator`.

### CODE REVIEW

File: app/Community/Controllers/ForumTopicController.php [hunk 1]
```diff
@@ -91,4 +97,41 @@ public function destroy(ForumTopic $topic): void
     {
         $this->authorize('delete', $topic);
     }
+
+    public function recentlyActive(): InertiaResponse
+    {
+        $recentlyActiveTopics = ForumTopic::query()
+            ->select('ForumTopic.*', 'ftc24h.comment_count as comment_count_24h', 'ftc24h.oldest_comment_id as oldest_comment_id_24h', 'ftc7d.comment_count as comment_count_7d', 'ftc7d.oldest_comment_id as oldest_comment_id_7d')
+            ->with(['latestComment.user'])
+            ->leftJoinSub($this->buildCommentCountSubQuery('24 HOUR'), 'ftc24h', 'ForumTopic.ID', '=', 'ftc24h.ForumTopicID')
```
Comment (wescopeland) [inline app/Community/Controllers/ForumTopicController.php:106]: The back-end code to build the page begins here. It works like this:
* Make a query to the DB.
* Map the query results to a Resource.
* Return props as paginated data from `LengthAwarePaginator`.

File: app/Data/PaginatedData.php
Comment (wescopeland) [inline app/Data/PaginatedData.php:1]: This resource receives a `LengthAwarePaginator` instance and packages it in a way that the Inertia+React code can understand it.

File: app/Data/Transformers/SiteEnumTransformer.php
Comment (wescopeland) [inline app/Data/Transformers/SiteEnumTransformer.php:1]: We have a lot of `abstract class` enums in the code. I whipped this up so those could automatically be converted to TypeScript enums. See: Platform/Enums/AchievementFlag.php.

File: app/Support/Shortcode/Shortcode.php [hunk 2]
```diff
@@ -113,8 +113,8 @@ public static function stripAndClamp(string $input, int $previewLength = 100): s
 
         // If the string is over the preview length, clamp it and add "..."
         // This can happen as a result of the replacement from above.
-        if (strlen($input) > $previewLength) {
-            $input = substr($input, 0, $previewLength) . '...';
+        if (mb_strlen($input) > $previewLength) {
+            $input = mb_substr($input, 0, $previewLength) . '...';
```
Comment (wescopeland) [inline app/Support/Shortcode/Shortcode.php:117]: This fixes a bug that caused JSON decoding to crash.

File: resources/js/types/generated.d.ts
Comment (wescopeland) [inline resources/js/types/generated.d.ts:1]: These are the auto-generated types that are spit out from the PHP `#[TypeScript]` annotations. You can create them yourself with the `composer types` command.

File: resources/js/types/ziggy.d.ts
Comment (wescopeland) [inline resources/js/types/ziggy.d.ts:1]: This is also auto-generated code.

**Comment by wescopeland** [comment]
@RetroAchievements/web-maintainers I've rebased this branch with the latest `master` to get rid of all the noise in the diff.

The PR is still large, but not 4000 lines large 😅 The diff now represents only the follow-up to #2502.

**Comment by Jamiras** [comment]
Not sure exactly what's going on, but the new implementation is twice as slow, uses three times as many queries, and loads 400x as many models.

master:
image
image

this branch:
image
image

**Comment by wescopeland** [comment]
>Not sure exactly what's going on, but the new implementation is twice as slow, uses three times as many queries, and loads 400x as many models.

Good catch. I underestimated how much the Eloquent ORM query was doing under the hood.

I believe if we're to maintain the same level of performance using Eloquent ORM, we'll essentially be writing raw SQL anyway.

With that in mind, in https://github.com/RetroAchievements/RAWeb/pull/2579/commits/571357a9a7859b847c0817c36bf1a9b4b99788f0 I've scaled things back a little bit to a happy medium. The original query is now being used as a private method in `ForumTopicController`, but the result is still mapped to a `LengthAwarePaginator`.

File: resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx [hunk 3]
```diff
@@ -10,37 +10,38 @@ import { PostTimestamp } from '../PostTimestamp';
...
-                href={`/viewtopic.php?t=${recentForumPost.forumTopicId}&c=${recentForumPost.commentId}#${recentForumPost.commentId}`}
+                href={`/viewtopic.php?t=${topic.id}&c=${topic.latestComment.id}#${topic.latestComment.id}`}
               >
-                {recentForumPost.forumTopicTitle}
+                {topic.title}
               </a>
             </p>
 
-            <p className="line-clamp-3 text-xs">{recentForumPost.shortMessage}</p>
+            <p className="line-clamp-3 text-xs">{topic.latestComment.body}</p>
```
Comment (Jamiras) [inline resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx:44]: The "N posts in the last 24 hours" / "N posts in the last week" are missing.

master:
image

this branch:
image

Comment (wescopeland) [reply resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx:44]: Good catch, thanks. When I changed the query around I deleted an `->include()` function call that I forgot to re-add. Fixed in latest.

File: app/Community/Controllers/ForumTopicController.php [hunk 4]
```diff
@@ -91,4 +98,110 @@ public function destroy(ForumTopic $topic): void
...
+        $count = 25;
+        $offset = ($page - 1) * $count;
+
+        $query = "
+            SELECT ft.ID as ForumTopicID, ft.Title as ForumTopicTitle,
+                   f.ID as ForumID, f.Title as ForumTitle,
+                   lc.CommentID, lftc.DateCreated as PostedAt, lftc.author_id,
+                   ua.User AS Author, ua.display_name AS AuthorDisplayName,
+                   LEFT(lftc.Payload, 90) AS ShortMsg,
+                   LENGTH(lftc.Payload) > 90 AS IsTruncated,
```
Comment (Jamiras) [inline app/Community/Controllers/ForumTopicController.php:172]: This limit seems significantly lower than the previous limit.

master:
image

this branch:
image

also the "..." appears to be missing.

File: resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx (hunk 3 above)
Comment (Jamiras) [reply resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx:44]: Last 24 hours is showing up now, but still not seeing last 7 days.

master:
image

this branch:
image

Comment (wescopeland) [reply resources/js/features/forums/components/RecentPostsMainRoot/RecentPostsCards/RecentPostsCards.tsx:44]: Looks like there was a logic error in the React code. Fixed in https://github.com/RetroAchievements/RAWeb/pull/2579/commits/8512598cec3d478a4a66bc17ad48302eccc44b1a and added another test case.

File: app/Community/Controllers/ForumTopicController.php (hunk 4 above)
Comment (wescopeland) [reply app/Community/Controllers/ForumTopicController.php]: Fixed in latest -- when I migrated the query to the controller I didn't notice we were overriding the default length of 90 with a different value.

Screenshot 2024-08-17 at 7 43 55 AM

**Review by Jamiras — APPROVED** [review]

**Pull request merged**